### PR TITLE
DEEP-9999: Using retryable http client for fetching license URLs

### PR DESCRIPTION
**Pull request by hrithik-777**
DEEP-9999: Using retryable http client for fetching license URLs

### CODE REVIEW

File: pkg/crawler/crawler.go [hunk 1]
```diff
@@ -508,7 +541,7 @@ func (c *Crawler) prepareClassifierData(ctx context.Context) ([]licenseFilesMeta
 
 				licenseFileName := getLicenseFileName(c.licensedir, licenseKey)
 				licenseMeta := uniqLicenseKeyMap[licenseKey]
-				ok, err := c.generateLicenseFile(client, licenseFileName, licenseMeta)
+				ok, err := GenerateLicenseFile(c.http, licenseFileName, licenseMeta)
```
Comment (theakshaymenon) [inline pkg/crawler/crawler.go:527]: why was this changed?

File: pkg/crawler/crawler.go [hunk 2]
```diff
@@ -586,24 +611,53 @@ func (c *Crawler) generateLicenseFile(client http.Client, licenseFileName string
...
 	}
-	if err != nil {
+	if resp.StatusCode != http.StatusOK {
+		log.Printf("Error while fetching license Meta URL %s, status code %d", licenseMeta.URL, resp.StatusCode)
 		return false, nil
 	}
 	defer resp.Body.Close()
 
-	_, err = io.Copy(f, resp.Body)
+	responseBody, err := HandleCompressedResponsebody(resp)
```
Comment (theakshaymenon) [inline pkg/crawler/crawler.go]: does this need to be exported `HandleCompressedResponsebody `?

**Review by theakshaymenon — APPROVED** [review]

File: pkg/crawler/crawler.go [hunk 3]
```diff
@@ -462,30 +471,54 @@ func (c *Crawler) prepareClassifierData(ctx context.Context) ([]licenseFilesMeta
...
+		req.Header.Set("Accept-Language", "en-US,en;q=0.5")
+		req.Header.Set("Accept-Encoding", "gzip, deflate, br")
+		req.Header.Set("Connection", "keep-alive")
+		req.Header.Set("Upgrade-Insecure-Requests", "1")
+		req.Header.Set("Cache-Control", "no-cache")
+	}
+
+	defer func() {
+		// reset the values set above
+		c.http.HTTPClient.Timeout = 0
```
Comment (namandf) [inline pkg/crawler/crawler.go]: Aren't we concurrently using the same crawler object? Won't this be a problem?

File: pkg/crawler/crawler.go [hunk 4]
```diff
@@ -586,24 +594,72 @@ func (c *Crawler) generateLicenseFile(client http.Client, licenseFileName string
...
+	case "br":
+		reader = brotli.NewReader(reader)
+
+	case "deflate":
+		deflateReader := flate.NewReader(reader)
+		defer deflateReader.Close()
+		reader = deflateReader
+	}
+
+	body, err := io.ReadAll(reader)
```
Comment (namandf) [inline pkg/crawler/crawler.go]: io.Copy is preferred for large files if i am not wrong, hence we were using io.Copy to a file earlier. It transfers data in parts.

File: pkg/crawler/crawler.go (hunk 3 above)
Comment (hrithik-777) [reply pkg/crawler/crawler.go]: Should not be a problem. the license classification is done at the end of Crawl API. Also regarding the license fetch in different go routines, we are closing the license channel first, next return, next the defer is executed. Do you think this can cause a problem? @namandf

Comment (namandf) [reply pkg/crawler/crawler.go]: But this isn't for classification. This is for preparing the classifier data i.e saving the files locally with license text. That happens as part of the crawl using goroutines. Right?

Comment (namandf) [reply pkg/crawler/crawler.go]: So if multiple goroutines are using the same crawler object and are updating the http client metadata, that might be a problem .

Comment (hrithik-777) [reply pkg/crawler/crawler.go:496]: No at the end of Crawl, we call the classify license. this then calls the prepareClassifierData. So this happens in main thread only.

File: pkg/crawler/crawler.go (hunk 4 above)
Comment (hrithik-777) [reply pkg/crawler/crawler.go]: got it. added it back
